- [ ] Importing basic libraries 
- [ ] Importing specific functions inside the python modules folder
- [ ] Defining the function transposonmapper 
    - **Importing data files**
        - [ ] Loading bam file from local data folder (IT IS NOT PART OF THE REPO FILES)
        - [ ] Loading gff file from data_file folder
        - [ ] loading text files with essential genes info in WT
        - [ ] loading text file with gene name aliases
    - **Reading files**
    - [ ] Read BAM file
    - **Heavy processing**
        - [ ] Get names of all chromosomes as stored in the bam file
        - [ ] Convert chromosome names in data file to roman numerals
        - [ ] Get sequence lengths of all chromosomes
        - [ ] Get number of mapped, unmapped and total amount of reads per chromosome
        - [ ] Get all reads within a specified genomic region 
            - [ ] retrieving all the reads from the current chromosome
            - [ ] get flag for each read. if read on forward strand , assign 1 , otherwise assign 0. 
            - [ ] correct starting position for reads with reverse orientation
            - [ ] create array of start position and flags for all reads in genome. 
        - [ ] get list of all genes and all essential genes (could be moved to **Importing data files**)
        - [ ] concatenate all chromosomes 
        - [ ] Get number of transposons and reads per gene (**critical step**)
    - **Writing output files** 
        - [ ] Create bed file
        - [ ] Create text file with transposons and reads per gene 
        - [ ] Create text file transposons and reads per essential gene 
        - [ ] Create text file with location of insertions and reads per gene 
        - [ ] Create text file with location of insertions and reads per essential gene
        - [ ] Add insertions at same location but with different orientations together (for the wig file) 
        - [ ] Create wig file 

