{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python376jvsc74a57bd0b2b6faf28e54db84b8bab82dacb459349e5fe486b652702d3c0dc9e6aff947a2",
   "display_name": "Python 3.7.6 64-bit ('base': conda)"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "## Importing basic libraries "
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os, sys\n",
    "import warnings\n",
    "import timeit\n",
    "import numpy as np\n",
    "import pysam"
   ]
  },
  {
   "source": [
    "## Importing existing python modules "
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dirname = os.path.dirname(os.path.abspath('__file__'))\n",
    "sys.path.insert(1,os.path.join(dirname,'..','python_scripts','python_modules'))\n",
    "from chromosome_and_gene_positions import chromosomename_roman_to_arabic, gene_position\n",
    "from gene_names import gene_aliases\n",
    "from samflag import samflags\n",
    "\n",
    "bam_arg = sys.argv[1]"
   ]
  },
  {
   "source": [
    "## Transposon mapping python script\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Defining the function transposonmapper \n",
    "\n",
    "#%%\n",
    "def transposonmapper(bamfile=bam_arg, gfffile=None, essentialfiles=None, genenamesfile=None):\n",
    "   \n",
    "\n",
    "### LOADING BAM FILE\n",
    "    if bamfile is None:\n",
    "        path = os.path.join('/data/localhome/linigodelacruz/Documents/PhD_2018/Documentation/SATAY/data/15022021-sequencing-data-WT-dnrp1-SATAY-from-Oxford/fastQ-data-files/align_out')\n",
    "        # filename = 'E-MTAB-4885.WT2.bam'\n",
    "        filename = 'WT_merged-DpnII-NlaIII-a_trimmed.sorted.bam'\n",
    "        bamfile = os.path.join(path,filename)\n",
    "    else:\n",
    "        filename = os.path.basename(bamfile)\n",
    "        path = bamfile.replace(filename,'')\n",
    "\n",
    "\n",
    "    assert os.path.isfile(bamfile), 'Bam file not found at: %s' % bamfile #check if given bam file exists\n",
    "\n",
    "\n",
    "### LOADING ADDITIONAL FILES from data_file folder\n",
    "    files_path = os.path.join(dirname,'..','data_files')\n",
    "\n",
    "    #LOADING GFF-FILE\n",
    "    if gfffile is None:\n",
    "        gfffile = os.path.join(files_path,'Saccharomyces_cerevisiae.R64-1-1.99.gff3')\n",
    "    assert os.path.isfile(gfffile), 'Path to GFF-file does not exist.'\n",
    "\n",
    "    #LOADING TEXT FILES WITH ESSENTIAL GENES\n",
    "    if essentialfiles is None:\n",
    "        essentialfiles = os.path.join(files_path,'Cerevisiae_AllEssentialGenes_List.txt')\n",
    "    assert os.path.isfile(essentialfiles), 'Following path does not exist: %s' % essentialfiles\n",
    "    del essentialfiles\n",
    "\n",
    "    #LOADING TEXT FILE WITH GENE NAME ALIASES\n",
    "    if genenamesfile is None:\n",
    "        genenamesfile = os.path.join(files_path,'Yeast_Protein_Names.txt')\n",
    "    assert os.path.isfile(genenamesfile), 'Following path does not exist: %s' % genenamesfile\n",
    "\n",
    "\n",
    "### READ BAM FILE\n",
    "    bam = pysam.AlignmentFile(bamfile, 'rb') #open bam formatted file for reading\n",
    "\n",
    "\n",
    "\n",
    "### GET NAMES OF ALL CHROMOSOMES AS STORED IN THE BAM FILE\n",
    "    ref_tid_dict = {} # 'I' | 0, 'II' | 1, ...\n",
    "    ref_name_list = [] # 'I', 'II', ...\n",
    "    for i in range(bam.nreferences): #if bam.nreferences does not work, use range(17) #16 chromosomes and the mitochondrial chromosome\n",
    "        ref_name = bam.get_reference_name(i)\n",
    "        ref_tid_dict[ref_name] = bam.get_tid(ref_name)\n",
    "        ref_name_list.append(ref_name)\n",
    "\n",
    "\n",
    "\n",
    "    del (ref_name, i)\n",
    "\n",
    "\n",
    "\n",
    "### CONVERT CHROMOSOME NAMES IN DATA FILE TO ROMAN NUMERALS\n",
    "    ref_romannums = chromosomename_roman_to_arabic()[0]\n",
    "    ref_tid_roman_dict = {}\n",
    "    for key,val in ref_tid_dict.items():\n",
    "        ref_tid_roman_dict[ref_romannums[int(val)+1]] = key\n",
    "\n",
    "\n",
    "\n",
    "    del (key, val, ref_romannums)\n",
    "\n",
    "\n",
    "\n",
    "### GET SEQUENCE LENGTHS OF ALL CHROMOSOMES\n",
    "    chr_length_dict = {} # 'I' | 230218, 'II' | 813184, ...\n",
    "    chr_summedlength_dict = {} # 'I' | 0, 'II' | 230218, 'III' |  1043402, ...\n",
    "    ref_summedlength = 0\n",
    "    for key in ref_tid_dict:\n",
    "        ref_length = bam.get_reference_length(key)\n",
    "        chr_length_dict[key] = ref_length\n",
    "        chr_summedlength_dict[key] = ref_summedlength\n",
    "        ref_summedlength += ref_length\n",
    "\n",
    "\n",
    "\n",
    "    del (key, ref_length, ref_summedlength)\n",
    "\n",
    "\n",
    "\n",
    "### GET NUMBER OF MAPPED, UNMAPPED AND TOTAL AMOUNT OF READS PER CHROMOSOME\n",
    "    # total_reads = bam.mapped\n",
    "    stats = bam.get_index_statistics()\n",
    "    chr_mappedreads_dict = {} # 'I' | [mapped, unmapped, total reads]\n",
    "    for stat in stats:\n",
    "        chr_mappedreads_dict[stat[0]] = [stat[1], stat[2], stat[3]]\n",
    "        if stat[2] != 0:\n",
    "            warnings.warn('Unmapped reads found in chromosome ' + stat[0])\n",
    "\n",
    "\n",
    "\n",
    "    del (stat, stats)\n",
    "\n",
    "\n",
    "\n",
    "### GET ALL READS WITHIN A SPECIFIED GENOMIC REGION\n",
    "    tnnumber_dict = {}\n",
    "    ll = 0 #Number of unique insertions in entire genome\n",
    "    for kk in ref_name_list:\n",
    "        timer_start = timeit.default_timer()\n",
    "        read_counter = 0\n",
    "\n",
    "        N_reads_kk = chr_mappedreads_dict[kk][2]\n",
    "        start_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "        flag_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "        readlength_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "\n",
    "        #RETREIVING ALL THE READS FROM THE CURRENT CHROMOSOME.\n",
    "        print('Getting reads for chromosome %s ...' % kk)\n",
    "        for reads in bam.fetch(kk, 0, chr_length_dict[kk], until_eof=True):\n",
    "            read = str(reads).split('\\t')\n",
    "\n",
    "            start_array[read_counter] = int(read[3]) + 1\n",
    "\n",
    "            #GET FLAG FOR EACH READ. IF READ ON FORWARD STRAND, ASSIGN VALUE 1, IF READ ON REVERSE STRAND ASSIGN VALUE -1, IF READ UNMAPPED OR SECONDARY ALIGNMENT ASSIGN VALUE 0\n",
    "#            flag_array[read_counter] = int(read[1])\n",
    "            samprop = samflags(flag = int(read[1]), verbose=False)[1]\n",
    "            if 'read reverse strand' in samprop:\n",
    "                flag_array[read_counter] = -1\n",
    "            else:\n",
    "                flag_array[read_counter] = 1\n",
    "            if 'not primary alignment' in samprop or 'read unmapped' in samprop:\n",
    "                flag_array[read_counter] = 0\n",
    "\n",
    "\n",
    "            cigarmatch_list = []\n",
    "            if not reads.cigartuples == None:\n",
    "                for cigar_type, cigar_length in reads.cigartuples:\n",
    "                    if cigar_type == 0:\n",
    "                        cigarmatch_list.append(cigar_length)\n",
    "                    elif cigar_type == 2:\n",
    "                        cigarmatch_list.append(cigar_length)\n",
    "            match_length = sum(cigarmatch_list)\n",
    "\n",
    "            readlength_array[read_counter] = match_length #int(len(read[9]))\n",
    "\n",
    "            read_counter += 1\n",
    "\n",
    "\n",
    "\n",
    "        #CORRECT STARTING POSITION FOR READS WITH REVERSED ORIENTATION\n",
    "#        flag0coor_array = np.where(flag_array==0) #coordinates reads 5' -> 3'\n",
    "#        flag16coor_array = np.where(flag_array==16) # coordinates reads 3' -> 5'\n",
    "        flag0coor_array = np.where(flag_array==1) #coordinates reads 5' -> 3'\n",
    "        flag16coor_array = np.where(flag_array==-1) # coordinates reads 3' -> 5'\n",
    "\n",
    "        startdirect_array = start_array[flag0coor_array]\n",
    "        flagdirect_array = flag_array[flag0coor_array]\n",
    "\n",
    "        startindirect_array = start_array[flag16coor_array] + readlength_array[flag16coor_array]\n",
    "        flagindirect_array = flag_array[flag16coor_array]\n",
    "\n",
    "        start2_array = np.concatenate((startdirect_array, startindirect_array), axis=0)\n",
    "        flag2_array = np.concatenate((flagdirect_array, flagindirect_array), axis=0)\n",
    "\n",
    "        del (flag0coor_array, flag16coor_array, startdirect_array, flagdirect_array, startindirect_array, flagindirect_array)\n",
    "\n",
    "\n",
    "\n",
    "        start2_sortindices = start2_array.argsort(kind='mergesort') #use mergesort for stable sorting\n",
    "        start2_array = start2_array[start2_sortindices]\n",
    "        flag2_array = flag2_array[start2_sortindices]\n",
    "\n",
    "        del start2_sortindices\n",
    "\n",
    "\n",
    "\n",
    "        #CREATE ARRAY OF START POSITION AND FLAGS OF ALL READS IN GENOME\n",
    "        ref_tid_kk = int(ref_tid_dict[kk]+1)\n",
    "        if ll == 0:\n",
    "            tncoordinates_array = np.array([])\n",
    "\n",
    "        mm = 0 # Number of unique reads per insertion\n",
    "        jj = 1 # Number of unique reads in current chromosome (Number of transposons in current chromosome)\n",
    "        for ii in range(1,len(start2_array)):\n",
    "            if abs(start2_array[ii]-start2_array[ii-1]) <= 2 and flag2_array[ii] == flag2_array[ii-1]: #If two subsequent reads are within two basepairs and have the same orientation, add them together.\n",
    "                mm += 1\n",
    "            else:\n",
    "                avg_start_pos = abs(round(np.mean(start2_array[ii-mm-1 : ii])))\n",
    "                if tncoordinates_array.size == 0: #include first read\n",
    "                    tncoordinates_array = np.array([ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])])\n",
    "                    readnumb_list = [mm+1]\n",
    "                else:\n",
    "                    tncoordinates_array = np.vstack((tncoordinates_array, [ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])]))    \n",
    "                    readnumb_list.append(mm+1)\n",
    "                mm = 0\n",
    "                jj += 1\n",
    "                ll += 1\n",
    "\n",
    "            if ii == len(start2_array) - 1: #include last read\n",
    "                avg_start_pos = abs(round(np.mean(start2_array[ii-mm-1 : ii])))\n",
    "                tncoordinates_array = np.vstack((tncoordinates_array, [ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])]))\n",
    "                readnumb_list.append(mm+1)\n",
    "\n",
    "        tnnumber_dict[kk] = jj\n",
    "\n",
    "        del (jj, start_array, flag_array, readlength_array, flag2_array, start2_array, ref_tid_kk)\n",
    "\n",
    "\n",
    "\n",
    "        timer_end = timeit.default_timer()\n",
    "        print('Chromosome %s completed in %.3f seconds' % (kk, (timer_end - timer_start)))\n",
    "        print('')\n",
    "\n",
    "\n",
    "\n",
    "    readnumb_array = np.array(readnumb_list)\n",
    "    del readnumb_list\n",
    "\n",
    "    tncoordinatescopy_array = np.array(tncoordinates_array, copy=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### GET LIST OF ALL GENES AND ALL ESSENTIAL GENES\n",
    "    print('Getting coordinates of all genes ...')\n",
    "\n",
    "    # GET POSITION GENES\n",
    "    gff_path = os.path.join(files_path,'Saccharomyces_cerevisiae.R64-1-1.99.gff3')\n",
    "    genecoordinates_dict = gene_position(gff_path) #'YAL069W' | ['I', 335, 649], ...\n",
    "\n",
    "\n",
    "\n",
    "    # GET ALL ANNOTATED ESSENTIAL GENES\n",
    "    essential_path = os.path.join(files_path,'Cerevisiae_AllEssentialGenes_List.txt')\n",
    "    essentialcoordinates_dict = {}\n",
    "    with open(essential_path, 'r') as f:\n",
    "        genes = f.readlines()[1:]\n",
    "        for gene in genes:\n",
    "            name = gene.strip('\\n')\n",
    "            essentialcoordinates_dict[name] = genecoordinates_dict.get(name).copy()\n",
    "\n",
    "\n",
    "    # GET ALIASES OF ALL GENES\n",
    "    names_path = os.path.join(files_path,'Yeast_Protein_Names.txt')\n",
    "    aliases_designation_dict = gene_aliases(names_path)[0] #'YMR056C' \\ ['AAC1'], ...\n",
    "\n",
    "\n",
    "\n",
    "    del (gff_path, gene, genes, name, essential_path)\n",
    "\n",
    "\n",
    "\n",
    "### CONCATENATE ALL CHROMOSOMES\n",
    "\n",
    "    #FOR EACH INSERTION LOCATION, ADD THE LENGTH OF ALL PREVIOUS CHROMOSOMES.\n",
    "    ll = 0\n",
    "    for ii in range(1,len(ref_name_list)):\n",
    "        ll += chr_length_dict[ref_name_list[ii-1]]\n",
    "        aa = np.where(tncoordinatescopy_array[:,0] == ii + 1)\n",
    "        tncoordinatescopy_array[aa,1] = tncoordinatescopy_array[aa,1] + ll\n",
    "\n",
    "\n",
    "\n",
    "    #FOR EACH GENE LOCATION, ADD THE LENGTH OF ALL PREVIOUS CHROMOSOMES.\n",
    "    for key in genecoordinates_dict:\n",
    "        gene_chrom = ref_tid_roman_dict.get(genecoordinates_dict.get(key)[0])\n",
    "        genecoordinates_dict[key][1] = genecoordinates_dict.get(key)[1] + chr_summedlength_dict.get(gene_chrom)\n",
    "        genecoordinates_dict[key][2] = genecoordinates_dict.get(key)[2] + chr_summedlength_dict.get(gene_chrom)\n",
    "\n",
    "\n",
    "\n",
    "    #FOR EACH ESSENTIAL GENE LOCATION, ADD THE LENGTH OF ALL PREVIOUS CHROMOSOMES.\n",
    "    for key in essentialcoordinates_dict:\n",
    "        gene_chrom = ref_tid_roman_dict.get(essentialcoordinates_dict.get(key)[0])\n",
    "        essentialcoordinates_dict[key][1] = essentialcoordinates_dict.get(key)[1] + chr_summedlength_dict.get(gene_chrom)\n",
    "        essentialcoordinates_dict[key][2] = essentialcoordinates_dict.get(key)[2] + chr_summedlength_dict.get(gene_chrom)\n",
    "\n",
    "\n",
    "\n",
    "    del (ii, ll, aa, key, gene_chrom)\n",
    "\n",
    "\n",
    "\n",
    "### GET NUMBER OF TRANSPOSONS AND READS PER GENE\n",
    "    print('Get number of insertions and reads per gene ...')\n",
    "    \n",
    "    #ALL GENES\n",
    "    tnpergene_dict = {}\n",
    "    readpergene_dict = {}\n",
    "    tncoordinates_pergene_dict = {}\n",
    "    # readpergenecrude_dict = {}\n",
    "    for gene in genecoordinates_dict:\n",
    "        xx = np.where(np.logical_and(tncoordinatescopy_array[:,1] >= genecoordinates_dict.get(gene)[1], tncoordinatescopy_array[:,1] <= genecoordinates_dict.get(gene)[2])) #get all insertions within range of current gene\n",
    "        tnpergene_dict[gene] = np.size(xx)\n",
    "        readpergene_dict[gene] = sum(readnumb_array[xx]) - max(readnumb_array[xx], default=0) #REMOVE LARGEST VALUE TO REDUCE NOISE\n",
    "        # readpergenecrude_dict[gene] = sum(readnumb_array[xx])\n",
    "\n",
    "        if np.size(xx) > 0:\n",
    "            tncoordinates_pergene_dict[gene] = [genecoordinates_dict.get(gene)[0], genecoordinates_dict.get(gene)[1], genecoordinates_dict.get(gene)[2], list(tncoordinatescopy_array[xx[0][0]:xx[0][-1]+1, 1]), list(readnumb_array[xx])]\n",
    "        else:\n",
    "            tncoordinates_pergene_dict[gene] = [genecoordinates_dict.get(gene)[0], genecoordinates_dict.get(gene)[1], genecoordinates_dict.get(gene)[2], [], []]\n",
    "\n",
    "\n",
    "    #ONLY ESSENTIAL GENES\n",
    "    tnperessential_dict = {}\n",
    "    readperessential_dict = {}\n",
    "    tncoordinates_peressential_dict = {}\n",
    "    # readperessentialcrude_dict = {}\n",
    "    for gene in essentialcoordinates_dict:\n",
    "        xx = np.where(np.logical_and(tncoordinatescopy_array[:,1] >= essentialcoordinates_dict.get(gene)[1], tncoordinatescopy_array[:,1] <= essentialcoordinates_dict.get(gene)[2]))\n",
    "        tnperessential_dict[gene] = np.size(xx)\n",
    "        readperessential_dict[gene] = sum(readnumb_array[xx]) - max(readnumb_array[xx], default=0)\n",
    "        # readperessentialcrude_dict[gene] = sum(readnumb_array[xx])\n",
    "\n",
    "        if np.size(xx) > 0:\n",
    "            tncoordinates_peressential_dict[gene] = [essentialcoordinates_dict.get(gene)[0], essentialcoordinates_dict.get(gene)[1], essentialcoordinates_dict.get(gene)[2], list(tncoordinatescopy_array[xx[0][0]:xx[0][-1]+1, 1]), list(readnumb_array[xx])]\n",
    "        else:\n",
    "            tncoordinates_peressential_dict[gene] = [essentialcoordinates_dict.get(gene)[0], essentialcoordinates_dict.get(gene)[1], essentialcoordinates_dict.get(gene)[2], [], []]\n",
    "\n",
    "\n",
    "\n",
    "    del (xx, gene)\n",
    "\n",
    "\n",
    "\n",
    "###CREATE BED FILE\n",
    "    bedfile = bamfile+'.bed'\n",
    "    print('Writing bed file at: ', bedfile)\n",
    "    print('')\n",
    "\n",
    "\n",
    "    with open(bedfile, 'w') as f:\n",
    "        \n",
    "        f.write('track name=' + filename + ' useScore=1\\n')\n",
    "        \n",
    "        coordinates_counter = 0\n",
    "        for tn in tncoordinates_array:\n",
    "            refname = [key for key, val in ref_tid_dict.items() if val == tn[0] - 1][0]\n",
    "            if refname == 'Mito':\n",
    "                refname = 'M'\n",
    "            f.write('chr' + refname + ' ' + str(tn[1]) + ' ' + str(tn[1] + 1) + ' . ' + str(100+readnumb_array[coordinates_counter]*20) + '\\n')\n",
    "            coordinates_counter += 1\n",
    "\n",
    "\n",
    "\n",
    "    del (bedfile, coordinates_counter, refname)\n",
    "\n",
    "### CREATE TEXT FILE WITH TRANSPOSONS AND READS PER GENE\n",
    "# NOTE THAT THE TRANSPOSON WITH THE HIGHEST READ COUNT IS IGNORED.\n",
    "# E.G. IF THIS FILE IS COMPARED WITH THE _PERGENE_INSERTIONS.TXT FILE THE READS DON'T ADD UP (SEE https://groups.google.com/forum/#!category-topic/satayusers/bioinformatics/uaTpKsmgU6Q)\n",
    "# TOO REMOVE THIS HACK, CHANGE THE INITIALIZATION OF THE VARIABLE readpergene\n",
    "    pergenefile = bamfile+'_pergene.txt'\n",
    "    print('Writing pergene.txt file at: ', pergenefile)\n",
    "    print('')\n",
    "\n",
    "    with open(pergenefile, 'w') as f:\n",
    "\n",
    "        f.write('Gene name\\tNumber of transposons per gene\\tNumber of reads per gene\\n')\n",
    "\n",
    "        for gene in tnpergene_dict:\n",
    "            tnpergene = tnpergene_dict[gene]\n",
    "            readpergene = readpergene_dict[gene]\n",
    "            if gene in aliases_designation_dict:\n",
    "                gene_alias = aliases_designation_dict.get(gene)[0]\n",
    "            else:\n",
    "                gene_alias = gene\n",
    "            f.write(gene_alias + '\\t' + str(tnpergene) + '\\t' + str(readpergene) + '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    del (pergenefile, gene, gene_alias, tnpergene, readpergene)\n",
    "\n",
    "### CREATE TEXT FILE TRANSPOSONS AND READS PER ESSENTIAL GENE\n",
    "    peressentialfile = bamfile+'_peressential.txt'\n",
    "    print('Writing peressential.txt file at: ',peressentialfile)\n",
    "    print('')\n",
    "\n",
    "    with open(peressentialfile, 'w') as f:\n",
    "        \n",
    "        f.write('Gene name\\tNumber of transposons per gene\\tNumber of reads per gene\\n')\n",
    "        \n",
    "        for essential in tnperessential_dict:\n",
    "            tnperessential = tnperessential_dict[essential]\n",
    "            readperessential = readperessential_dict[essential]\n",
    "            if essential in aliases_designation_dict:\n",
    "                essential_alias = aliases_designation_dict.get(essential)[0]\n",
    "            else:\n",
    "                essential_alias = essential\n",
    "            f.write(essential_alias + '\\t' + str(tnperessential) + '\\t' + str(readperessential) + '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    del (peressentialfile, essential, essential_alias, tnperessential, readperessential)\n",
    "\n",
    "\n",
    "### CREATE TEXT FILE WITH LOCATION OF INSERTIONS AND READS PER GENE\n",
    "    pergeneinsertionsfile = bamfile+'_pergene_insertions.txt'\n",
    "    print('Witing pergene_insertions.txt file at: ',pergeneinsertionsfile)\n",
    "    print('')\n",
    "\n",
    "    with open(pergeneinsertionsfile, 'w') as f:\n",
    "\n",
    "        f.write('Gene name\\tChromosome\\tStart location\\tEnd location\\tInsertion locations\\tReads per insertion location\\n')\n",
    "\n",
    "        for gene in tncoordinates_pergene_dict:\n",
    "            gene_chrom = ref_tid_roman_dict.get(genecoordinates_dict.get(gene)[0])\n",
    "            tncoordinates = [ins - chr_summedlength_dict.get(gene_chrom) for ins in tncoordinates_pergene_dict[gene][3]]\n",
    "\n",
    "            if gene in aliases_designation_dict:\n",
    "                gene_alias = aliases_designation_dict.get(gene)[0]\n",
    "            else:\n",
    "                gene_alias = gene\n",
    "\n",
    "            f.write(gene_alias + '\\t' + str(tncoordinates_pergene_dict[gene][0]) + '\\t' + str(tncoordinates_pergene_dict[gene][1] - chr_summedlength_dict.get(gene_chrom)) + '\\t' + str(tncoordinates_pergene_dict[gene][2] - chr_summedlength_dict.get(gene_chrom)) + '\\t' + str(tncoordinates) + '\\t' + str(tncoordinates_pergene_dict[gene][4]) + '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    del (gene, gene_chrom, tncoordinates, gene_alias, pergeneinsertionsfile)\n",
    "\n",
    "### CREATE TEXT FILE WITH LOCATION OF INSERTIONS AND READS PER ESSENTIAL GENE\n",
    "    peressentialinsertionsfile = bamfile+'_peressential_insertions.txt'\n",
    "    print('Writing peressential_insertions.txt file at: ', peressentialinsertionsfile)\n",
    "    print('')\n",
    "\n",
    "    with open(peressentialinsertionsfile, 'w') as f:\n",
    "\n",
    "        f.write('Essential gene name\\tChromosome\\tStart location\\tEnd location\\tInsertion locations\\tReads per insertion location\\n')\n",
    "\n",
    "        for essential in tncoordinates_peressential_dict:\n",
    "            gene_chrom = ref_tid_roman_dict.get(genecoordinates_dict.get(essential)[0])\n",
    "            tncoordinates = [ins - chr_summedlength_dict.get(gene_chrom) for ins in tncoordinates_peressential_dict[essential][3]]\n",
    "\n",
    "            if essential in aliases_designation_dict:\n",
    "                essential_alias = aliases_designation_dict.get(essential)[0]\n",
    "            else:\n",
    "                essential_alias = essential\n",
    "\n",
    "            f.write(essential_alias + '\\t' + str(tncoordinates_peressential_dict[essential][0]) + '\\t' + str(tncoordinates_peressential_dict[essential][1] - chr_summedlength_dict.get(gene_chrom)) + '\\t' + str(tncoordinates_peressential_dict[essential][2] - chr_summedlength_dict.get(gene_chrom)) + '\\t' + str(tncoordinates) + '\\t' + str(tncoordinates_peressential_dict[essential][4]) + '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    del (essential, gene_chrom, tncoordinates, essential_alias, peressentialinsertionsfile)\n",
    "\n",
    "### ADD INSERTIONS AT SAME LOCATION BUT WITH DIFFERENT ORIENTATIONS TOGETHER (FOR STORING IN WIG-FILE)\n",
    "    wigfile = bamfile+'.wig'\n",
    "    print('Writing wig file at: ', wigfile)\n",
    "    print('')\n",
    "\n",
    "\n",
    "    readnumbwig_array = readnumb_array.copy()\n",
    "\n",
    "\n",
    "\n",
    "    unique_index_array = np.array([], dtype=int) #=cc\n",
    "    N_uniques_perchr_list = []\n",
    "    ll = 0\n",
    "    for kk in ref_name_list:\n",
    "        index = np.where(tncoordinates_array[:,0] == int(ref_tid_dict[kk]+1)) #get indices for current chromosome.\n",
    "        unique_index = np.unique(tncoordinates_array[index][:,1], return_index=True)[1] #get all insertion locations (in tncoordinates, all rows, column 1)\n",
    "\n",
    "        unique_index_array = np.append(unique_index_array, (unique_index+ll), axis=0)\n",
    "\n",
    "        ll += np.count_nonzero(tncoordinates_array[:,0] == int(ref_tid_dict[kk]+1))\n",
    "        N_uniques_perchr_list.append(ll) #total amount unique indices found untill current chromosome\n",
    "\n",
    "\n",
    "\n",
    "    del (ll, kk, unique_index)\n",
    "\n",
    "\n",
    "\n",
    "    duplicate_list = [] #=dd\n",
    "    ll = 0\n",
    "    index_last_unique_previous_chromosome = 0\n",
    "    for ii in N_uniques_perchr_list:\n",
    "        index_last_unique = np.where(unique_index_array <= ii)[0][-1]\n",
    "        for jj in range(ll,ii):\n",
    "            if int(jj) not in unique_index_array[index_last_unique_previous_chromosome:index_last_unique]:\n",
    "                duplicate_list.append(jj)\n",
    "        index_last_unique_previous_chromosome = index_last_unique\n",
    "        ll = ii\n",
    "\n",
    "    #SUM READNUMB VALUES AT INDEX IN DUPLICATE_LIST AND DUPLICATE_LIST-1    \n",
    "    for ii in duplicate_list:\n",
    "        readnumbwig_array[ii-1] = readnumbwig_array[ii-1] + readnumbwig_array[ii]\n",
    "    \n",
    "    tncoordinateswig_duplicatesremoved_array = np.delete(tncoordinates_array, duplicate_list, axis=0)\n",
    "    readnumbwig_duplicatesremoved_array = np.delete(readnumbwig_array, duplicate_list, axis=0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    del (ll, ii, jj, N_uniques_perchr_list, index_last_unique, duplicate_list, readnumbwig_array)\n",
    "\n",
    "### CREATING WIG FILE\n",
    "    with  open(wigfile, 'w') as f:\n",
    "        f.write('track type=wiggle_0 ,maxheightPixels=60 name='+filename+'\\n')\n",
    "        for kk in ref_name_list:\n",
    "            f.write('VariableStep chrom=chr' + kk + '\\n')\n",
    "\n",
    "            index = np.where(tncoordinateswig_duplicatesremoved_array[:,0] == int(ref_tid_dict[kk]+1)) #get indices for current chromosome.\n",
    "            for ii in index[0]:\n",
    "                f.write(str(tncoordinateswig_duplicatesremoved_array[ii][1]) + ' ' + str(readnumbwig_duplicatesremoved_array[ii]) + '\\n')\n",
    "\n",
    "\n",
    "    del (wigfile, kk, ii, index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "KeyError",
     "evalue": "17",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-1633c5eacc6f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtransposonmapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbamfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-16-d57dd1ebd9f1>\u001b[0m in \u001b[0;36mtransposonmapper\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m     62\u001b[0m     \u001b[0mref_tid_roman_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mref_tid_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m         \u001b[0mref_tid_roman_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mref_romannums\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 17"
     ]
    }
   ],
   "source": [
    "transposonmapper(bamfile=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}